codec/enforceproto/extraneous_fields_test.go

```diff
@@ -0,0 +1,258 @@
+		},
+		{
+			name: "Extra field that's non-existent in Customer1, along with Reserved set",
+			in: &testdata.Customer2{
+				Id:       289,
+				Name:     "Customer1",
+				Industry: 5299,
+				Fewer:    199.9,
+				Reserved: 819,
+			},
+			wantErr: &UnmatchedTagsError{
+				[]*tagAndWireType{{4, 5}},
+			},
+		},
+		{
+			name: "Extra field that's non-existent in Customer1, along with Reserved set",
+			in: &testdata.Customer2{
+				Id:       289,
+				Name:     "Customer1",
+				Industry: 5299,
+				Fewer:    199.9,
+				Reserved: 819,
+			},
+			wantErr: &UnmatchedTagsError{
+				[]*tagAndWireType{{4, 5}},
+			},
+		},
+		{
+			name: "Use enumerated field",
+			in: &testdata.Customer2{
+				Id:       289,
+				Name:     "Customer1",
+				Industry: 5299,
+				City:     testdata.Customer2_PaloAlto,
+			},
+			wantErr: &UnmatchedTagsError{
+				[]*tagAndWireType{{6, 0}},
+			},
+		},
+		{
+			name: "multiple extraneous fields",
+			in: &testdata.Customer2{
+				Id:       289,
+				Name:     "Customer1",
+				Industry: 5299,
+				City:     testdata.Customer2_PaloAlto,
+				Fewer:    45,
+			},
+			wantErr: &UnmatchedTagsError{
+				[]*tagAndWireType{
+					{4, 5},
+					{6, 0},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			blob, err := proto.Marshal(tt.in)
+			if err != nil {
+				t.Fatalf("Failed to marshal input: %v", err)
+			}
+
+			c1 := new(testdata.Customer1)
+			gotErr := CheckExtraneousFieldsProto(blob, c1)
+			if !reflect.DeepEqual(gotErr, tt.wantErr) {
+				t.Fatalf("Error mismatch\nGot:\n%s\n\nWant:\n%s", gotErr, tt.wantErr)
+			}
+		})
+	}
+}
+
+type satisfyProtoMessage struct {
+}
+
+var _ gogoproto.Message = (*satisfyProtoMessage)(nil)
+
+func (spm *satisfyProtoMessage) String() string { return "satisfyProtoMessage" }
+
+func (spm *satisfyProtoMessage) Reset()        {}
+func (spm *satisfyProtoMessage) ProtoMessage() {}
+
+type testT1 struct {
+	ID              int32   `protobuf:"varint,1,opt,name=id,proto3" json:"id,omitempty"`
+	Name            string  `protobuf:"bytes,2,opt,name=name,proto3" json:"name,omitempty"`
+	SubscriptionFee float32 `protobuf:"fixed32,3,opt,name=subscription_fee,json=subscriptionFee,proto3" json:"subscription_fee,omitempty"`
+	satisfyProtoMessage
+}
+
+type testT2 struct {
+	ID       int32                   `protobuf:"varint,1,opt,name=id,proto3" json:"id,omitempty"`
+	Industry int32                   `protobuf:"varint,2,opt,name=industry,proto3" json:"industry,omitempty"`
+	Name     string                  `protobuf:"bytes,3,opt,name=name,proto3" json:"name,omitempty"`
+	Fewer    float32                 `protobuf:"fixed32,4,opt,name=fewer,proto3" json:"fewer,omitempty"`
+	Reserved int64                   `protobuf:"varint,1047,opt,name=reserved,proto3" json:"reserved,omitempty"`
+	City     testdata.Customer2_City `protobuf:"varint,6,opt,name=city,proto3,enum=service.Customer2_City" json:"city,omitempty"`
+	satisfyProtoMessage
+}
```

Comment:
I'm not sure why we need to manually create test structs. These should be created using the some `make proto-gen` as the rest of the SDK proto code.

Reply:
Yeah they were manually created because it has been about 2 years since I was last here so forgot how they Cosmos generated them but you helped me so this is done :)